communote/webapp/src/main/webapp/javascript/utils/EventEmitter.js

```diff
@@ -0,0 +1,98 @@
+(function() {
```

Comment:
Generell könnte man auf eigene EventListener verzichten und lieber auf Standard-JS/jQuery gehen:

`document.addEventListener('cnt-widget/refreshed', ....)`

und im Widget dann

`document.dispatchEvent('cnt-widget/refreshed', ...)`

Das wäre so ein Ding, wo man ältere Browser (also IE <=8) konsequent ausschließen sollte.

Reply:
Darüber hab ich auch nachgedacht, finde das aber semantisch und architektonisch unsauber/falsch:
- document.dispatchEvent(String, ...) gibt es nicht. Das erste Argument ist eine Instanz des Event Interfaces.
- ich möchte meine Events nicht auf einen DOM Knoten feuern, sondern auf ein spezielles Event-Target wie z.B. ein Widget oder irgend eine bel. andere Komponente, die keine Repräsentation im DOM haben muss.
  - Es ist auch sinnvoll, dass EventListener auf einer bestimmten Komponente mit aufgeräumt werden, wenn die Komponente nicht mehr existiert. Das hat man bei der document basierten Variante nicht.
- in JavaScript gibt es dafür das EventTarget Interface (wird z.B. von Element oder dem XMLHttpRequest implementiert). Müsste man aber auch selber implementieren und kann dann wieder nur Instanzen vom Event Interface dispatchen. Das Interface hat aber auch einen sehr starken DOM Bezug und für "normale" Events zu viele unnötige Member.
- Alternative wäre das Mootools Event Mixin gewesen. Sehr elegant, aber von Mootols will ich eher weg.
- Daher hab ich den schlanken EventEmitter gewählt, den es in ähnlicher Form auch im NodeJS gibt. Damit hat man volle Flexibilität. Idealerweise bauen wir die anderen Stellen, an denen wir auch noch Event-Targets implementiert haben, auf den EventEmitter um (Konsistenz).